src/htm/algorithms/TemporalMemory.cpp

```diff
@@ -603,16 +604,15 @@ SDR TemporalMemory::getPredictiveCells() const {
   correctDims.push_back(static_cast<CellIdx>(getCellsPerColumn()));
   SDR predictive(correctDims);
 
-  auto& predictiveCells = predictive.getSparse();
+  std::unordered_set<CellIdx> uniqueCells;
+  uniqueCells.reserve(activeSegments_.size());
 
-  for (auto segment = activeSegments_.cbegin(); segment != activeSegments_.cend();
-       segment++) {
-    const CellIdx cell = connections.cellForSegment(*segment);
-    if (segment == activeSegments_.begin() || cell != predictiveCells.back()) {
-      predictiveCells.push_back(cell);
-    }
+  for (const auto segment : activeSegments_) {
+    const CellIdx cell = connections.cellForSegment(segment);
+    uniqueCells.insert(cell); //set keeps the cells unique
```

Comment:
FYI @Zbysekz the former code was a hack that exploited the fact that cells come sorted. Took me a while to understand that too, the new code is much more readable and I believe likely even faster (haven't microbenchmarked)